src/pytorch/pme/pme.py

```diff
@@ -0,0 +1,194 @@
+from ..neighbors import getNeighborPairs
+import torch
+import math
+
+class PME:
+    """This class implements the Particle Mesh Ewald algorithm (https://doi.org/10.1063/1.470117).
+
+    This is a method of summing all the infinite pairwise electrostatic interactions in a periodic system.  It divides
+    the energy into two parts: a short range term that can be computed efficiently in direct space, and a long range
+    term that can be computed efficiently in reciprocal space.  The individual terms are not physical meaningful, only
+    their sum.
+
+    This class supports periodic boundary conditions with arbitrary triclinic boxes.  The box vectors `a`, `b`, and `c`
+    must satisfy certain requirements:
+
+    `a[1] = a[2] = b[2] = 0`
+    `a[0] >= 2*cutoff, b[1] >= 2*cutoff, c[2] >= 2*cutoff`
+    `a[0] >= 2*b[0]`
+    `a[0] >= 2*c[0]`
+    `b[1] >= 2*c[1]`
+
+    These requirements correspond to a particular rotation of the system and reduced form of the vectors, as well as the
+    requirement that the cutoff be no larger than half the box width.
+
+    You can optionally specify that certain interactions should be omitted when computing the energy.  This is typically
+    used for nearby atoms within the same molecule.  When two atoms are listed as an exclusion, only the interaction of
+    each with the same periodic copy of the other (that is, not applying periodic boundary conditions) is excluded.
+    Each atom still interacts with all the periodic copies of the other.
+
+    Due to the way the reciprocal space term is calculated, it is impossible to prevent it from including excluded
+    interactions.  The direct space term therefore compensates for it, subtracting off the energy that was incorrectly
+    included in reciprocal space.  The sum of the two terms thus yields the correct energy with the interaction fully
+    excluded.
+
+    When performing backpropagation, this class computes derivatives with respect to atomic positions and charges, but
```

Comment:
Maybe I am over amplifying this because it made me waste a lot of time recently, but I believe there should be a mention here to this module having only a single backwards pass.
If one trains with forces this will be grad'ed during the forward pass, which requires second grad during backprop.

Reply:
Yeah, that's an annoying limitation.  It shouldn't be an issue for the sorts of problems I'm interested in, since you wouldn't use PME during training.  You would be training on isolated (not periodic) small molecules, for which all interactions are computed in direct space.  PME would only be used for inference, if you want to apply the model to larger systems.  It would be more of an issues for materials applications, if you wanted to train on crystals.

I'll see if I can get it to throw an exception if you try to take a second derivative.

Reply:
Python-side there is the @once_differentiable decorator: https://discuss.pytorch.org/t/what-does-the-function-wrapper-once-differentiable-do/31513

Reply:
It turns out that it already throws an exception:

```
RuntimeError: element 0 of tensors does not require grad and does not have a grad_fn
```

I added a test case to check for it.